pallets/automation-time/src/lib.rs

```diff
@@ -15,26 +15,186 @@
+use sp_runtime::traits::SaturatedConversion;
+use sp_std::vec;
 
 #[frame_support::pallet]
 pub mod pallet {
+	use super::*;
+
+	type AccountOf<T> = <T as frame_system::Config>::AccountId;
+
+	#[derive(Debug, Eq, PartialEq, Encode, Decode, TypeInfo)]
+	#[scale_info(skip_type_params(T))]
+	pub enum Action {
+		Notify(Vec<u8>),
+	}
+
+	#[derive(Debug, Eq, PartialEq, Encode, Decode, TypeInfo)]
+	#[scale_info(skip_type_params(T))]
+	pub struct Task<T: Config> {
+		owner_id: AccountOf<T>,
+		time: u64,
+		action: Action,
+	}
+
+	impl<T: Config> Task<T> {
+		pub fn create_event_task(owner_id: AccountOf<T>, time: u64, message: Vec<u8>) -> Task<T> {
+			let action = Action::Notify(message);
+			Task::<T> { owner_id, time, action }
+		}
+	}
+
 	#[pallet::config]
-	pub trait Config: frame_system::Config {}
+	pub trait Config: frame_system::Config + pallet_timestamp::Config {
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// The maximum number of tasks that can be scheduled for a time slot.
+		#[pallet::constant]
+		type MaxTasksPerSlot: Get<u32>;
+	}
 
 	#[pallet::pallet]
 	#[pallet::generate_store(pub(super) trait Store)]
 	pub struct Pallet<T>(_);
 
+	#[pallet::storage]
+	#[pallet::getter(fn get_scheduled_tasks)]
+	pub type ScheduledTasks<T: Config> =
+		StorageMap<_, Twox64Concat, u64, BoundedVec<T::Hash, T::MaxTasksPerSlot>>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn get_task)]
+	pub type Tasks<T: Config> = StorageMap<_, Twox64Concat, T::Hash, Task<T>>;
+
 	#[pallet::error]
 	pub enum Error<T> {
-		Placeholder,
+		/// Time must end in a whole minute.
+		InvalidTime,
+		/// Time must be in the future.
+		PastTime,
+		/// The message cannot be empty.
+		EmptyMessage,
+		/// There can be no duplicate tasks.
+		DuplicateTask,
+		/// Time slot is full. No more tasks can be scheduled for this time.
+		TimeSlotFull,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Schedule task success. [who, task_id]
+		TaskScheduled(T::AccountId, T::Hash),
 	}
 
 	#[pallet::call]
-	impl<T: Config> Pallet<T> {}
+	impl<T: Config> Pallet<T> {
+		/// Schedule a task to fire an event with a custom message.
+		///
+		/// Before the task can be scheduled the task must past validation checks.
+		/// * The transaction is signed
+		/// * The time is valid
+		/// * The message's length > 0
+		///
+		/// # Parameters
+		/// * `time`: The unix standard time in seconds for when the task should run.
+		/// * `message`: The message you want the event to have.
+		///
+		/// # Errors
+		/// * `InvalidTime`: Time must end in a whole minute.
+		/// * `PastTime`: Time must be in the future.
+		/// * `EmptyMessage`: The message cannot be empty.
+		/// * `DuplicateTask`: There can be no duplicate tasks.
+		/// * `TimeSlotFull`: Time slot is full. No more tasks can be scheduled for this time.
+		#[pallet::weight(10_000 + T::DbWeight::get().writes(2) + T::DbWeight::get().reads(2))]
```

Comment:
how'd we calculate this weight?

Reply:
really its a placeholder. I will get the real values in phase 4. Notice that this pallet isn't even exposed to the runtime